### PR TITLE
Added test for MGF, added missing checks in crv_types

### PR DESCRIPTION
**Pull request by ritesh99rakesh**
#### References to other Issues or PRs

Closes #16571 

#### Brief description of what is fixed or changed
Added tests for `moment_generating_function` for continuous distributions.
Added missing checks and property `set` for crv_types.
Speed up tests in `test_continuous_rv` by replacing `simplify` with `expand` where ever possible.
Removed `@XFAIL` for all but one test.

#### Other comments
Under GSoC'19

#### Release Notes

- stats
    - Added missing checks and property `set` for crv_types
    - Corrected `_moment_generating_function` for `Gumbel` distribution

### CODE REVIEW

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal Please review this.

**Comment by czgdp1807** [comment]
@ritesh99rakesh Please write, Closes https://github.com/sympy/sympy/pull/16571 in the references section of your first comment.

File: sympy/stats/crv_types.py [hunk 1]
```diff
@@ -163,6 +163,8 @@ def rv(symbol, cls, args):
 class ArcsinDistribution(SingleContinuousDistribution):
     _argnames = ('a', 'b')
 
+    set = Interval(0, 1)
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: I think it should be `set = Interval(a, b)`as given [here](https://en.wikipedia.org/wiki/Arcsine_distribution#Generalization).

File: sympy/stats/crv_types.py [hunk 2]
```diff
@@ -1765,6 +1785,12 @@ def Kumaraswamy(name, a, b):
 class LaplaceDistribution(SingleContinuousDistribution):
     _argnames = ('mu', 'b')
 
+    set = Interval(-oo, oo)
+
+    @staticmethod
+    def check(mu, b):
+        _value_check(b > 0, "Scale parameter b must be positive.")
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: What about `_value_check(mu.is_real, "Location parameter mu should be real")`?

File: sympy/stats/crv_types.py [hunk 3]
```diff
@@ -2770,6 +2825,19 @@ def StudentT(name, nu):
...
     _argnames = ('a', 'b', 'c', 'd')
 
+    @property
+    def set(self):
+        return Interval(self.a, self.d)
+
+    @staticmethod
+    def check(a, b, c, d):
+        _value_check(a < d, "Lower bound parameter a < %s. a = %s"%(d, a))
+        _value_check(And(a <= b, b < c),
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: According to latest refactoring of `_value_check` ,`And(a <= b, b < c)` should be written as, `(a <= b, b < c)`.

Comment (czgdp1807) [reply sympy/stats/crv_types.py]: > According to latest refactoring of `_value_check` ,`And(a <= b, b < c)` should be written as, `(a <= b, b < c)`.

+1

File: sympy/stats/crv_types.py [hunk 4]
```diff
@@ -2770,6 +2825,19 @@ def StudentT(name, nu):
...
+        "Level start parameter b must be in range [%s, %s). b = %s"%(a, c, b))
+        _value_check(And(b < c, c <= d),
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: According to latest refactoring of `_value_check`, `And(b < c, c <= d)` should be written as, `(b < c, c <= d)`.

File: sympy/stats/crv_types.py [hunk 5]
```diff
@@ -2850,6 +2918,16 @@ def Trapezoidal(name, a, b, c, d):
...
     _argnames = ('a', 'b', 'c')
 
+    @property
+    def set(self):
+        return Interval(self.a, self.b)
+
+    @staticmethod
+    def check(a, b, c):
+        _value_check(b > a, "Parameter b > %s. b = %s"%(a, b))
+        _value_check(And(a <= c, c <= b),
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: According to latest refactoring of `_value_check`, `And(a <= c, c <= b)` should be written as, `(a <= c, c <= b)`.

File: sympy/stats/crv_types.py [hunk 6]
```diff
@@ -3047,6 +3133,11 @@ class UniformSumDistribution(SingleContinuousDistribution):
     def set(self):
         return Interval(0, self.n)
 
+    @staticmethod
+    def check(n):
+        _value_check(And(n > 0, n.is_integer),
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: Write, `(n > 0, n.is_integer)`.

File: sympy/stats/tests/test_continuous_rv.py [hunk 7]
```diff
@@ -795,22 +859,18 @@ def test_input_value_assertions():
...
+    "(2*sqrt(pi)), (x, -oo, oo))")
 
-    assert E(X + 1, evaluate=False) == (
-        Integral(sqrt(2)*x*exp(-x**2/2)/(2*sqrt(pi)), (x, -oo, oo)) + 1)
+    assert str(E(X + 1, evaluate=False)) == ("Integral(sqrt(2)*x*exp(-x**2/2)/"
+    "(2*sqrt(pi)), (x, -oo, oo)) + 1")
 
-    assert P(X > 0, evaluate=False) == (
-        Integral(sqrt(2)*exp(-x**2/2)/(2*sqrt(pi)), (x, 0, oo)))
+    assert str(P(X > 0, evaluate=False)) == ("Integral(sqrt(2)*exp(-_z**2/2)/"
```
Comment (czgdp1807) [inline sympy/stats/tests/test_continuous_rv.py]: What's the output without using `str`?

**Comment by sidhantnagpal** [comment]
The release notes entry can be skipped, unless it is a user-facing change.

File: sympy/stats/tests/test_continuous_rv.py [hunk 8]
```diff
@@ -147,12 +151,76 @@ def test_characteristic_function():
...
+    assert cf(1) == exp(I - S(1)/2)
 
     Z = Exponential('z', 5)
     cf = characteristic_function(Z)
     assert cf(0) == 1
-    assert simplify(cf(1)) == S(25)/26 + 5*I/26
+    assert cf(1).expand() == S(25)/26 + 5*I/26
+
+def test_moment_generating_function():
+    t = S('t')
```
Comment (czgdp1807) [inline sympy/stats/tests/test_continuous_rv.py]: This should be declared with appropriate assumptions. As far as I know, `t` should be real, so apply that assumption.

Comment (supreet11agrawal) [reply sympy/stats/tests/test_continuous_rv.py]: I agree with @czgdp1807. Also, I don't think that tests such as these, check if the `mgf` generated is correct or not. perhaps, you can check for actual expressions on Wikipedia or Wolfram and check if the function is generating them.

**Review by supreet11agrawal — CHANGES_REQUESTED** [review]

**Comment by czgdp1807** [comment]
Please add symbolic tests wherever possible. Numeric tests are only good when the results being generated are unevaluated and too long.

File: sympy/stats/crv_types.py (hunk 1 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Thank you!

File: sympy/stats/tests/test_continuous_rv.py (hunk 7 above)
Comment (ritesh99rakesh) [reply sympy/stats/tests/test_continuous_rv.py]: Same, but it does not evaluate to true while testing, the reason being the use of dummy `_z`.  @sidhantnagpal suggested this [here](https://github.com/sympy/sympy/pull/16827#discussion_r283788020).

File: sympy/stats/crv_types.py (hunk 6 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Done. Thanks!

File: sympy/stats/crv_types.py (hunk 5 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Done. Thanks!

File: sympy/stats/crv_types.py (hunk 4 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Done. Thanks!

File: sympy/stats/crv_types.py (hunk 3 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Done. Thanks!

File: sympy/stats/crv_types.py (hunk 2 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: Done. Thanks!

**Comment by ritesh99rakesh** [comment]
@Abdullahjavednesar I have updated the commit, please remove the `Author's Turn` label. Thanks!

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal please review this.

File: sympy/stats/crv_types.py [hunk 9]
```diff
@@ -1205,6 +1210,13 @@ def FDistribution(name, d1, d2):
 class FisherZDistribution(SingleContinuousDistribution):
     _argnames = ('d1', 'd2')
 
+    set = Interval(-oo, oo)
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: Should this be `Interval(0, oo)`?

File: sympy/stats/crv_types.py [hunk 10]
```diff
@@ -2085,6 +2123,11 @@ class NakagamiDistribution(SingleContinuousDistribution):
 
     set = Interval(0, oo)
 
+    @staticmethod
+    def check(mu, omega):
+        _value_check(mu >= 0.5, "Shape parameter mu must be positive.")
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: `mu >= S.Half`. The message should be modified to fit the check.

File: sympy/stats/tests/test_continuous_rv.py [hunk 11]
```diff
@@ -147,12 +151,144 @@ def test_characteristic_function():
...
+    assert mgf == (1 - t/b)**(-a)
+
+    mgf = moment_generating_function(Exponential('x', a))(t)
+    assert mgf == a/(a - t)
+
+    mgf = moment_generating_function(Gamma('x', a, b))(t)
+    assert mgf == (-b*t + 1)**(-a)
+
+    mgf = moment_generating_function(Gumbel('x', a, b))(t)
+    assert mgf == exp(I*b*t)*gamma(-a*t + 1)
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_continuous_rv.py]: Shouldn't the exponential term have been `exp(b*t)`?

Comment (czgdp1807) [reply sympy/stats/tests/test_continuous_rv.py]: AFAIK, the PDF of Gumbel Distribution hasn't been corrected in master. It was merged in the release branch. Am I right @sidhantnagpal?

File: sympy/stats/tests/test_continuous_rv.py [hunk 12]
```diff
@@ -147,12 +151,144 @@ def test_characteristic_function():
...
+
+    mgf = moment_generating_function(RaisedCosine('x', a, b))(t)
+    assert mgf == pi**2*exp(a*t)*sinh(b*t)/(b*t*(b**2*t**2 + pi**2))
+
+    mgf = moment_generating_function(Rayleigh('x', a))(t)
+    assert mgf == sqrt(2)*sqrt(pi)*a*t*(erf(sqrt(2)*a*t/2) + 1)\
+        *exp(a**2*t**2/2)/2 + 1
+
+    mgf = moment_generating_function(Triangular('x', a, b, c))(t)
+    assert str(mgf) == ("(-2*(-a + b)*exp(c*t) + 2*(a + c)*exp(b*t) + "
```
Comment (sidhantnagpal) [inline sympy/stats/tests/test_continuous_rv.py]: Shouldn't the coefficient of `exp(b*t)` be `2*(-a + c)`?

**Comment by sidhantnagpal** [comment]
I have reviewed the changes and left some comments, overall it seems good to me. Thanks for fixing the `@XFAIL` tests.

It would be good to add tests for the staticmethod check added to the distributions, in the cases they will raise. Those can be dealt in a different PR. Also, the check methods in `joint_rv_types` distributions could also use the `@staticmethod` decorator.

File: sympy/stats/tests/test_continuous_rv.py (hunk 11 above)
Comment (sidhantnagpal) [reply sympy/stats/tests/test_continuous_rv.py]: > PDF of Gumbel Distribution hasn't been corrected in master

It has the release branches are merged back into master after the release. You can see the version of the method in the branch (and even check the git history to see that the pdf of Gumbel was changed).
https://github.com/sympy/sympy/blob/6f7d053881472f426cf13e6793d782f43b4afa9f/sympy/stats/crv_types.py#L1575-L1578

The issue seems to be in the definition of `Gumbel._moment_generating_function`:
https://github.com/sympy/sympy/blob/6f7d053881472f426cf13e6793d782f43b4afa9f/sympy/stats/crv_types.py#L1587-L1588

Comment (sidhantnagpal) [reply sympy/stats/tests/test_continuous_rv.py]: > PDF of Gumbel Distribution hasn't been corrected in master

It has, the release branches are merged back into master after the release. You can see the version of the method in the branch (and even check the git history to see that the pdf of Gumbel was changed).
https://github.com/sympy/sympy/blob/6f7d053881472f426cf13e6793d782f43b4afa9f/sympy/stats/crv_types.py#L1575-L1578

The issue seems to be in the definition of `Gumbel._moment_generating_function`:
https://github.com/sympy/sympy/blob/6f7d053881472f426cf13e6793d782f43b4afa9f/sympy/stats/crv_types.py#L1587-L1588

File: sympy/stats/crv_types.py (hunk 9 above)
Comment (ritesh99rakesh) [reply sympy/stats/crv_types.py]: The wiki link: https://en.wikipedia.org/wiki/Fisher's_z-distribution
says that `set = Interval(-oo, oo)`

Comment (sidhantnagpal) [reply sympy/stats/crv_types.py]: Ah, looking it up directed to "Fisher Distribution" https://en.wikipedia.org/wiki/F-distribution.

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal I have made the required changes. I too, thought that the tests could be added in another PR, preferably in the end. Please review.

**Comment by sidhantnagpal** [comment]
Did you look into failures? You can also test it locally using `bin/test stats` and `bin/doctest stats` to check if the tests pass for stats module.

**Comment by sidhantnagpal** [comment]
Looks good. The remaining tests for `_characteristic_function` and `check` can be added in a different PR.

**Pull request merged**